Add exhaustion to damage source

parlough opened the pull request:
[SpongeAPI](https://github.com/SpongePowered/SpongeAPI/pull/1674) | **SpongeCommon**

Adds the ability to set and retrieve the exhaustion of a damage source. Also adds a test plugin and fixes some small comments. I'm open to changes of the defaults for exhaustion and where I set the exhaustion when building the damage sources.

gabizou commented on testplugins/src/main/java/org/spongepowered/test/DamageSourceTest.java:
```diff
@@ -0,0 +1,109 @@
+
+    @Listener(order = Order.POST)
+    public void onPlayerDamage(DamageEntityEvent event, @Getter("getTargetEntity") Player player) {
+        event.getCause().first(DamageSource.class).ifPresent(d -> {
```
Could do with another filter instead of a nested lambda.

parlough replied:
Addressed.

gabizou commented on testplugins/src/main/java/org/spongepowered/test/DamageSourceTest.java:
```diff
@@ -0,0 +1,109 @@
+        Sponge.getCommandManager().register(this,
+                CommandSpec.builder()
+                        .executor((src, args) -> {
+                            if (src instanceof Player) {
```
If non player command sources aren’t valid, should return a specific type of return, right @dualspiral?

parlough replied:
Correct. I just put this test together quickly I guess. Will clean it up and address both issues.

parlough replied:
Issue is addressed.

dualspiral replied:
Yeah, an exception should be thrown, as it is now.

gabizou reviewed (commented):
Overall looks clean. Just a brief review from the phone. Will dive deeper in a bit.

parlough committed:
Adds exhaustion to damage sources and does some general cleanup

parlough committed:
Address various concerns and update for API changes

parlough commented:
So we've been calculating damage wrong in some cases and there there was a problem where we calculated enchantment damage modifiers even if the damage was absolute, so I added a check there. We were also not calculating armor damage modifiers if the damage was absolute when we should have been, so I removed that check and replaced it with the proper check to see if the damage source is bypassing armor. 

Other concerns have been addressed as well, most were in the API.

parlough commented on src/main/java/org/spongepowered/common/event/damage/DamageEventHandler.java:
```diff
@@ -227,78 +227,80 @@ public static EquipmentType resolveEquipment(int slot) {
-            NBTTagList enchantmentList = itemStack.getEnchantmentTagList();
-            if (enchantmentList == null) {
-                continue;
+        if (!damageSource.isDamageAbsolute()) {
```
No idea why this sections looks like it has huge differences in different places. This was the only change I made, obviously indenting everything inside.

Meronat commented:
Will update this once I get a chance soon!

parlough commented:
I will update this soon(hopefully tonight or tomorrow), most likely splitting out the fixes to our damage handling. Cheers.

Dockter commented:
Hope to see this in API 7

parlough commented:
@Dockter I'll try to get it done tonight then!
